pandas/core/indexes/datetimelike.py

```diff
@@ -637,13 +638,32 @@ def _sub_datelike(self, other):
     def _sub_period(self, other):
         return NotImplemented
 
-    def _add_offset_array(self, other):
-        # Array/Index of DateOffset objects
-        return NotImplemented
+    def _addsub_offset_array(self, other, op):
+        """
+        Add or subtract array-like of DateOffset objects
 
-    def _sub_offset_array(self, other):
-        # Array/Index of DateOffset objects
-        return NotImplemented
+        Parameters
+        ----------
+        other : Index, np.ndarray
+            object-dtype containing pd.DateOffset objects
+        op : {operator.add, operator.sub}
+
+        Returns
+        -------
+        result : same class as self
+        """
+        if len(other) == 1:
+            return op(self, other[0])
+
+        warnings.warn("Adding/subtracting array of DateOffsets to "
+                      "{cls} not vectorized"
+                      .format(cls=type(self).__name__), PerformanceWarning)
+
+        res_values = op(self.astype('O').values, np.array(other))
+        kwargs = {}
+        if not is_period_dtype(self):
+            kwargs['freq'] = 'infer'
+        return self.__class__(res_values, **kwargs)
```

Comment:
use ``.shallow_copy``

Reply:
Changed it to shallow_copy and it broke several tests, so just pushed with it changed back to `self.__class__`.  For some reason `DatetimeIndex._simple_new` returns an `ndarray` if its input values have object-dtype, which is the case here.

Reply:
so try using ``_shallow_copy_with_infer``, we by definition *never* *ever* use ``self.__class__`` in Index construction.